This is very cool, unfortunately I'm getting an error when I try to build this:
```
error[E0599]: `Workspaces` is not an iterator
   --> src/modules/hyprland.rs:38:30
    |
38  |   ...                   .into_iter()
    |                          ^^^^^^^^^ `Workspaces` is not an iterator
    |
   ::: /home/b3nj4m1n/.local/share/cargo/git/checkouts/hyprland-rs-ce7ad3b93f8820bb/9c92cd0/src/data/regular.rs:209:1
    |
209 | / create_data_struct!(
210 | |     vec Workspaces,
211 | |     DataCommands::Workspaces,
212 | |     Workspace,
213 | |     "This type provides a vector of workspaces"
214 | | );
    | | -
    | | |
    | |_doesn't satisfy `Workspaces: IntoIterator`
    |   doesn't satisfy `Workspaces: std::iter::Iterator`
    |
    = note: the following trait bounds were not satisfied:
            `Workspaces: std::iter::Iterator`
            which is required by `Workspaces: IntoIterator`
            `&Workspaces: std::iter::Iterator`
            which is required by `&Workspaces: IntoIterator`
            `&mut Workspaces: std::iter::Iterator`
            which is required by `&mut Workspaces: IntoIterator`

error[E0609]: no field `data` on type `&_`
  --> src/modules/hyprland.rs:45:67
   |
45 |                         wspaces.sort_by_key(|wspace| match wspace.data.id {
   |                                                                   ^^^^

Some errors have detailed explanations: E0599, E0609.
For more information about an error, try `rustc --explain E0599`.
```

I just added a bit of CI to this project in adfcd51907a3ddc5be7d74f3eab9895de83e14dc, so it'd be great if you could update your branch to include that commit, since I think that's required to run the GitHub action on this PR.

Once you do that, we'll know if this is an issue on my end or if there's actually a problem with the code.

---

> later will add logging

My plan was to add [tracing](https://lib.rs/crates/tracing) eventually.

---

Unfortunately, I haven't gotten around to trying out hyprland yet because I (sadly) have an nvidia card, so I'll have to figure out some way to actually test this PR once I get it to build, and I'm not sure when I'll be able to do that, just as a heads-up.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -20,9 +20,11 @@ mpd = { version = "0.0.12", optional = true }
 time = { version = "0.1.44", optional = true }
 pulse = { version = "2.0", package = "libpulse-binding", optional = true }
 sway = { version = "3.0.1", package = "swayipc", optional = true }
+hyprland= {git = 'https://github.com/hyprland-community/hyprland-rs.git', optional = true}
 
 [features]
-default = ["mpd", "pulseaudio", "sway"]
+default = ["mpd", "pulseaudio", "sway", "hyprland"]
 mpd = ["dep:mpd", "dep:time"]
 pulseaudio = ["dep:pulse"]
 sway = ["dep:sway"]
+hyprland = ["dep:hyprland"]

--- a/src/main.rs
+++ b/src/main.rs
@@ -4,7 +4,7 @@ use std::process::exit;
 use clap::{ColorChoice, Parser, Subcommand};
 
 mod modules;
-use modules::{mpd, pulseaudio, sway};
+use modules::{hyprland, mpd, pulseaudio, sway};
 use serde::Serialize;
 
 #[derive(Parser)]
@@ -41,6 +41,8 @@ enum Modules {
     PulseAudio,
     #[command(alias = "i3")]
     Sway,
+    #[command(subcommand)]
+    Hyprland(hyprland::HyprlandOpts),
 }
 
 #[derive(Debug, Serialize)]
@@ -99,6 +101,14 @@ fn main() {
             Modules::Sway => {
                 if cfg!(feature = "sway") {
                     while let Err(..) = (sway::Sway {}.start(5)) {}
+                    exit(0);
+                } else {
+                    println!("Feature not enabled");
+                }
+            }
+            Modules::Hyprland(ref opts) => {
+                if cfg!(feature = "hyprland") {
+                    while let Err(..) = hyprland::HyprlandListener::new(opts).listen() {}
                     exit(0);
                 } else {
                     println!("Feature not enabled");

--- a/src/modules/hyprland.rs
+++ b/src/modules/hyprland.rs
@@ -1,0 +1,127 @@
+use anyhow::Result;
+use clap::Subcommand;
+use hyprland::{
+    data::{Client, Devices, Workspace, Workspaces},
+    event_listener::EventListener,
+    prelude::*,
+};
+use serde::Serialize;
+
+#[derive(Subcommand)]
+pub enum HyprlandOpts {
+    Workspace,
+    Window,
+    Keyboard,
+}
+
+#[derive(Serialize)]
+struct WorkspaceData {
+    is_active: bool,
+    data: Workspace,
+}
+
+pub struct HyprlandListener {
+    listener: EventListener,
+}
+
+#[derive(Debug, Serialize)]
+struct KeyboardLayout(hyprland::data::Keyboard);
+impl HyprlandListener {
+    pub fn new(opts: &HyprlandOpts) -> Self {
+        let mut listener = EventListener::new();
+        match opts {
+            HyprlandOpts::Workspace => {
+                let print_workspace = || {
+                    if let Ok(wspaces) = Workspaces::get() {
+                        let active_workspace = Workspace::get_active();
+                        let mut wspaces: Vec<_> = wspaces
+                            .into_iter()
+                            .map(|w| {
+                                let is_active =
+                                    matches!(&active_workspace, Ok(space) if space.id == w.id);
+                                WorkspaceData { is_active, data: w }
+                            })
+                            .collect();
+                        wspaces.sort_by_key(|wspace| match wspace.data.id {
+                            hyprland::shared::WorkspaceType::Unnamed(id) => id,
+                            hyprland::shared::WorkspaceType::Named(_) => i32::MAX,
+                            hyprland::shared::WorkspaceType::Special(_) => i32::MAX,
+                        });
+                        crate::print(&Some(wspaces));
+                    } else {
+                        crate::print::<()>(&None);
+                    }
+                };
+                // for initial;
+                print_workspace();
+                listener.add_workspace_added_handler(move |wtype| {
+                    eprintln!("Workspace {wtype:?} added");
+                    print_workspace()
+                });
+                listener.add_workspace_moved_handler(move |mon_event| {
+                    eprintln!("Moniter changed: {mon_event:?}");
+                    print_workspace()
+                });
+                listener.add_workspace_change_handler(move |wtype| {
+                    eprintln!("Workspace {wtype:?} changed");
+                    print_workspace()
+                });
+                listener.add_workspace_destroy_handler(move |wtype| {
+                    eprintln!("Workspace {wtype:?} removed");
+                    print_workspace()
+                });
+                listener.add_fullscreen_state_change_handler(move |_state| {
+                    print_workspace();
+                });
+            }
+            HyprlandOpts::Window => {
+                let print_window = || {
+                    if let Ok(Some(client)) = Client::get_active() {
+                        crate::print(&Some(client.class));
+                    } else {
+                        crate::print::<()>(&None);
+                    }
+                };
+                listener.add_window_open_handler(move |win_event| {
+                    eprintln!("Active window opened: {win_event:?}");
+                    print_window();
+                });
+                listener.add_window_close_handler(move |win_event| {
+                    eprintln!("Window closed {win_event:?}");
+                    print_window();
+                });
+                listener.add_window_moved_handler(move |win_event| {
+                    eprintln!("Window moved: {win_event:?}");
+                    print_window();
+                });
+                listener.add_active_window_change_handler(move |win_event| {
+                    eprintln!("Window changed: {win_event:?}");
+                    print_window();
+                });
+            }
+            HyprlandOpts::Keyboard => {
+                let print_keyboard =
+                    || {
+                        if let Ok(devs) = Devices::get() {
+                            crate::print(&Some(devs.keyboards.get(0).expect(
+                                "It is assumed that Atleast one keyboard should be present",
+                            )));
+                        } else {
+                            crate::print::<()>(&None);
+                        };
+                    };
+                // while in the beginning prinnt the active keyboard
+                print_keyboard();
+                listener.add_keyboard_layout_change_handler(move |layout| {
+                    eprintln!("KeyboardLayout changed: {layout:?}");
+                    print_keyboard();
+                })
+            }
+        };
+        Self { listener }
+    }
+    pub fn listen(self) -> Result<()> {
+        self.listener.start_listener()?;
+        Ok(())
+    }
+}

--- a/src/modules/mod.rs
+++ b/src/modules/mod.rs
@@ -1,3 +1,4 @@
+pub mod hyprland;
 pub mod mpd;
 pub mod pulseaudio;
 pub mod sway;

